File: Hash/HashTablePractice.py
# Practice 12. Use hash tables
from itertools import combinations
TWO_SUM = 'T'
SYMMETRIC_PAIRS = 'S'

def readIntegers(n: int, s: str) -> list:
  words = s.split()
  assert(len(words) == n)
  arr = [int(w) for w in words]
  return arr

def twoSum(n, arr):
  hashTable = {}
  for i, j in combinations(arr, 2):
    hashTable[i+j] = True
  return hashTable.get(n)

def symmetricPairs(arr):
  hashTable = {}
  symmetric = []
  for pair in arr:
    exist = hashTable.get(pair[0]) and hashTable.get(pair[1])
    if exist:
      symmetric.append(pair)
    else:
      hashTable[pair[0]] = pair[1]
      hashTable[pair[1]] = pair[0]
  symmetric = list(map(lambda x: (x[1], x[0]) if x[0] > x[1] else x, symmetric))
  return symmetric

if __name__ == "__main__":
  with open('input.txt', 'r') as inFile:
    lines = inFile.readlines()
  with open('output.txt', 'w') as outFile:
    i = 0
    while i < len(lines):
      words = lines[i].split()
      op = words[0]

      if op == TWO_SUM:
        if len(words) != 3:
          raise Exception("TWO_SUM: invalid input")
        k, n = int(words[1]), int(words[2])
        i += 1
        nums = readIntegers(n, lines[i])
        summed = twoSum(k, nums)
        if summed:
          outFile.write("T\n")
        else:
          outFile.write("F\n")
        # TODO 
        # Take k and nums as input, and run twoSum(k, nums)
        # Write whether two sum exists into output file

      elif op == SYMMETRIC_PAIRS:
        if len(words) != 2:
          raise Exception("SYMMETRIC_PAIRS: invalid input")
        n = int(words[1])
        pairs = []
        while n:
          i += 1
          words = lines[i].split()
          if len(words) != 2:
            raise Exception("SYMMETRIC_PAIRS: invalid input")
          a, b = int(words[0]), int(words[1])
          pairs.append((a, b))
          n -= 1
        symPairs = symmetricPairs(pairs)
        for pair in symPairs:
          outFile.write(f"{pair[0]} {pair[1]}\n")
        # TODO
        # Take pairs as input, and run symmetricPairs(pairs)
        # Write every symmetric pair into output file

      else:
        raise Exception("Undefined operator")
      i += 1

File: Hash/HashTablePractice.cpp
// Practice 12. Use hash tables
#include <fstream>
#include <iostream>
#include <sstream>
#include <unordered_map>
#include <vector>
#include <algorithm>
using namespace std;
const char TWO_SUM = 'T';
const char SYMMETRIC_PAIRS = 'S';

vector<int> readIntegers(int k, string& line) {
  vector<int> arr;
  istringstream liss(line);
  int val;
  for (int i = 0; i < k; ++i) {
    if (!(liss >> val)) {
      cerr<<"Error: invalid input value"<<endl;
      exit(1);
    }
    else {
      arr.push_back(val);
    }
  }
  return arr;
}

bool twoSum(int n, vector<int> arr) {
  unordered_map<int, bool> hashTable;
  for (int i=0; i<arr.size(); i++)
    for (int j=i; j<arr.size(); j++)
        hashTable[arr[i] + arr[j]] = true;
  return hashTable[n];
}

vector<pair<int, int> > symmetricPairs(vector<pair<int, int> > arr) {
  unordered_map<int, int> hashTable;
  vector<pair<int, int> > symmetric;
  for (vector<pair<int, int> >::iterator iter=arr.begin(); iter!=arr.end(); iter++) {
    if ((hashTable[iter->first] == iter->second) && hashTable[iter->second] == iter->first)
      symmetric.push_back(*iter);
    else {
      hashTable[iter->first] = iter->second;
      hashTable[iter->second] = iter->first;
    }
  }
  for (int i=0; i<symmetric.size(); i++)
    swap(symmetric[i].first, symmetric[i].second);
  sort(symmetric.begin(), symmetric.end());
  return symmetric;
}

int main() {
  ifstream inFile("input.txt");
  ofstream outFile("output.txt");
  string line;
  while (getline(inFile, line))
  {
    char op = line[0];
    istringstream iss(line.substr(2));
    vector<int> nums;
    vector<pair<int, int> > pairs;
    vector<pair<int, int> > ans;
    vector<pair<int, int> >::iterator it;
    int k, n, a, b;
    bool exist;
    switch(op)
    {
      case TWO_SUM:
        if (!(iss >> k >> n)) {
          cerr<<"TWO_SUM: invalid input"<<endl;
          exit(1);
        }
        if (!getline(inFile, line)) {
          cerr<<"TWO_SUM: unable to read a line"<<endl;
          exit(1);
        }
        nums = readIntegers(n, line);
        exist = twoSum(k, nums);
        if (exist)
          outFile << "T" << endl;
        else
          outFile << "F" << endl;
        break;

      case SYMMETRIC_PAIRS:
        if (!(iss >> n)) {
          cerr<<"SYMMETRIC_PAIRS: invalid input"<<endl;
          exit(1);
        }
        for (int i = 0; i < n; ++i) {
          if (!getline(inFile, line)) {
            cerr<<"SYMMERIC_PAIRS: invalid input"<<endl;
            exit(1);
          }
          istringstream tiss(line);
          if (!(tiss >> a >> b)) {
            cerr<<"SYMMETRIC_PAIRS: invalid input"<<endl;
            exit(1);
          }
          pairs.push_back(make_pair(a, b));
        }
        ans = symmetricPairs(pairs);
        for (vector<pair<int, int> >::iterator iter=ans.begin(); iter!=ans.end(); iter++)
          outFile << iter->first << " " << iter->second << endl;
        // TODO
        // Take pairs as input, and run symmetricPairs(pairs)
        // Write every symmetric pair into output file
        break;

      default:
        cerr<<"Undefined operator"<<endl;
        exit(1);
    }
  }
  outFile.close();
  inFile.close();
}
